Crypto/AsyncCrypto.cpp

```diff
@@ -0,0 +1,350 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check it.
+
+// PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+#include "pch.h"
+#include "AsyncCrypto.h"
+
+AsyncCrypto::AsyncCrypto()
+{
+    m_keypair = std::make_shared<EVP_PKEY*>();
+    m_rsa_encr_ctx = std::make_shared<EVP_CIPHER_CTX*>();
+    m_rsa_decr_ctx = std::make_shared<EVP_CIPHER_CTX*>();
+
+    InitializeContext();
+
+    GenerateRsaKeypair();
+}
+
+AsyncCrypto::~AsyncCrypto()
+{
+    DestroyContext();
+}
+
+bool AsyncCrypto::InitializeContext()
+{
+    *m_rsa_encr_ctx.get() = EVP_CIPHER_CTX_new();
+    *m_rsa_decr_ctx.get() = EVP_CIPHER_CTX_new();
+
+    if (!*m_rsa_encr_ctx.get() || !*m_rsa_decr_ctx.get()) {
+        return false;
+    }
+    return true;
+}
+
+void AsyncCrypto::DestroyContext()
+{
+    EVP_PKEY_free(*m_keypair.get());
+
+    EVP_CIPHER_CTX_free(*m_rsa_encr_ctx.get());
+    EVP_CIPHER_CTX_free(*m_rsa_decr_ctx.get());
+}
+
+void AsyncCrypto::NotifyError(std::string& msg)
+{
+    throw exceptionAsyncCrypto(msg);
+}
+
+bool AsyncCrypto::GenerateRsaKeypair()
+{
+    EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, nullptr);
+
+    if (EVP_PKEY_keygen_init(ctx) <= 0) {
+        return false;
+    }
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, RSA_KEY_LEN) <= 0) {
+        return false;
+    }
+    if (EVP_PKEY_keygen(ctx, m_keypair.get()) <= 0) {
+        return false;
+    }
+
+    EVP_PKEY_CTX_free(ctx);
+
+    return true;
+}
+
+unsigned int AsyncCrypto::Encrypt(
+    const unsigned char* msg,
+    unsigned int msg_len,
+    std::shared_ptr<unsigned char>& encr_msg,
+    std::shared_ptr<unsigned char>& encr_key,
+    unsigned int* encr_key_len,
+    std::shared_ptr<unsigned char>& iv,
+    unsigned int* iv_len)
+{
+    if (!msg || !msg_len)
+    {
+        std::string msg = "message is empty";
+        NotifyError(msg);
```

Comment:
We are going to use crypto lib in server and server runs as a windows service and it is bad idea to use exception in windows service. You are not able to see this exception in windows services